### ISSUE TITLE
Incorrect file list maintenance causes compiler error in particular cases

### ISSUE THREAD

**Issue by oryol** [TypeStrong/ts-loader#71]
Prerequisites:
 * There are several TypeScript files in the project (as separated external modules because I want to use multiple bundles)
 * There are some npm based dependencies

The flow which causes to error:
 * When fist file are processed by ts-loader - typescript tries to load dependent npm modules. It checks package.json of the appropriate module (it can contain "typings" with direct reference to .d.ts).
 * TypeScript calls moduleResolutionHost.fileExists and readFile. And it's processed (in the ts-loader) through servicesHost.getScriptSnapshot (which is incorrect I think). As result - package.json is added to files array.
 * When next file is loaded using ts-loader (but in the same webpack instance of course) - TypeScript compiler calls serviceHost.getScriptFileNames and it returns package.json files which causes build fail without any understandable error, only "TypeScript emitted no output", because ts-loader doesn't show error messages from getCompilerOptionsDiagnostics and there is no error for specific source file, just for input file list (it contains json files).

As a temporary fix (obiously, it's better to decouple module resolution host from the service host, it's just a workaround) I fixed ts-loader/index.js (directly in the my node_modules) using following code in the getScriptFileNames:

```
return Object.keys(files).filter(function(f) { return !!f.match(/\.tsx?$/); });
```

It causes that json files are not returned for TypeScript input but still are saved in the files array (maybe it's ok to cache package.json in this array, I'm not sure actually).

Also I suggest to add getCompilerOptionsDiagnostics to error list in some way. Because it's very hard to understand what is causing error in some cases.

**Comment by jbrantly** [comment]
Thanks for the detailed report!

I think you're probably right about not using getScriptSnapshot for the moduleResolutionHost. It was done that way because that's how TS does it automatically if you don't override `resolveModuleNames` but since we are it probably doesn't make sense.

I'm not sure what you meant by ts-loader not showing error messages from `getCompilerOptionsDiagnostics` though. It should: https://github.com/TypeStrong/ts-loader/blob/master/index.ts#L350

**Comment by oryol** [comment]
I see. But it's invoked only when compiler is created. In my case the first file compiled successfully, but 2nd file compilation fails (actually, file list was wrong on this stage, it contained json files).

As result error messages says only:
```
Module build failed: Error: Typescript emitted no output for ...
    at Object.loader (...\node_modules\ts-loader\index.js:434:15)
```

After I have changed ```getScriptFileNames``` (added filtering by file extension) - everything compiled successfully. I have checked in the debugger - at least after ```languageService.getEmitOutput``` for the 2nd file (unfortunately I didn't discovered when exactly it appears) ```getCompilerOptionsDiagnostics``` returns error messages about json files.

If it's needed I can create minimal reproducible example for this case.

**Comment by jbrantly** [comment]
I don't think it's needed (yet). If I can't reproduce I'll let you know. What version of TypeScript are you using?

**Comment by oryol** [comment]
1.6.2

**Comment by alexgorbatchev** [comment]
I can confirm that proposed solution fixes the `TypeScript emitted no output` error.

**Comment by jbrantly** [comment]
I can't duplicate this after all :( I'm probably missing a setting somewhere. Could you provide a minimal example, including tsconfig.json and ideally webpack.config.js?

**Comment by jbrantly** [comment]
I put my attempt at duplicating this in #75 in case it helps. I basically took the existing `nodeResolution` test and added an additional dependency so that the loader was invoked twice (once for `app.ts` and once for `b.ts`).

I verified that `getScriptFileNames` was being called and that it did indeed contain the `package.json` file but I wasn't able to get an error.

**Comment by alexgorbatchev** [comment]
@jbrantly I emailed project sample to your public email.

**Comment by jbrantly** [comment]
Ah-ha! You need `noEmitOnError` enabled to duplicate. Thanks @alexgorbatchev

**Comment by alexgorbatchev** [comment]
Very cool @jbrantly, kudos for your amazing efforts! Thank you.

**Comment by oryol** [comment]
Good to know that it has been fixed!
Just to know - when are you going to release it to the *npm*?

**Comment by jbrantly** [comment]
In a day or two, probably. There are a couple other issues I'd like to get into the same release, and the PR still needs signing off.

**Comment by jbrantly** [comment]
This was published in v0.5.6.

Also, thanks @oryol and @alexgorbatchev for reporting this and helping debug/fix it, and apologies for the delay in getting it out.

**Comment by alexgorbatchev** [comment]
Fantastic, thank you so much!
